src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/service/SystemService.java

```diff
@@ -184,6 +190,82 @@ public AnchorFile getAnchorFile() throws AnchorNotFoundException {
         return anchorFile;
     }
 
+    /**
+     * Calculate the hex hash of the given anchor file. Used to verify/preview an anchor file before
+     * uploading it
+     * @param anchorBytes
+     * @return
+     * @throws InvalidAnchorInstanceException anchor is not generated in the current instance
+     */
+    public AnchorFile getAnchorFileFromBytes(byte[] anchorBytes) throws InvalidAnchorInstanceException {
+        ConfigurationAnchorV2 anchor = new ConfigurationAnchorV2(anchorBytes);
+        verifyAnchorInstance(anchor);
+        AnchorFile anchorFile = new AnchorFile(calculateAnchorHexHash(anchorBytes));
+        anchorFile.setCreatedAt(FormatUtils.fromDateToOffsetDateTime(anchor.getGeneratedAt()));
+        return anchorFile;
+    }
+
+    /**
+     * Upload a new configuration anchor. A temporary anchor file is created on the filesystem in order to run
+     * the verification process with configuration-client module (via external script).
+     * @param anchorBytes
+     * @throws InvalidAnchorInstanceException anchor is not generated in the current instance
+     * @throws AnchorUploadException in case of external process exceptions
+     */
+    public void uploadAnchor(byte[] anchorBytes) throws InvalidAnchorInstanceException, AnchorUploadException {
+        ConfigurationAnchorV2 anchor = new ConfigurationAnchorV2(anchorBytes);
+        verifyAnchorInstance(anchor);
+        File tempAnchor = null;
+        try {
+            tempAnchor = createTemporaryAnchorFile(anchorBytes);
+            configurationVerifier.verifyInternalConfiguration(tempAnchor.getAbsolutePath());
+            anchorRepository.saveAndReplace(tempAnchor);
+        } catch (ConfigurationVerifier.ConfigurationVerificationException ce) {
+            throw new DeviationAwareRuntimeException(ce, ce.getErrorDeviation());
+        } catch (CodedException ce) {
+            throw ce;
+        } catch (ServiceException | InterruptedException e) {
+            throw new AnchorUploadException(e);
```

Comment:
instead of

`} catch (ServiceException | InterruptedException e) {`

I think I would prefer

`} catch (ProcessNotExecutableException | ProcessFailedException | InterruptedException e) {`

- it is more descriptive, it explains which types of errors form an AnchorUploadException. ServiceException is quite abstract. OK, abstraction can intentional also, it could be argued otherwise too. I personally was slightly confused "where do these ServiceExceptions come from?" and would have preferred my suggestion.
- ServiceException is a bit of a catch-all, so when methods get modified, and new exceptions are added, we may mistakenly translate some new exception type to AnchorUploadException, even if it should have a different handling. So, I think it is _good_ that we force developer to rethink "is this AnchorUploadException or not" when new exceptions are added. This, as long as the number of discreet exception types does not get too big, and since it is only 3 at this point, it's not a problem

Reply:
Yeah I was deliberately going for the gotta-catch-'em-all approach since no matter what ServiceException we get here -> we still know that the anchor upload fails -> we know we want to throw AnchorUploadException. 

But I get the point you are saying here and it is probably better in the long run that we keep being verbose about the different exceptions as long as it doesn't get too overwhelming. Updated the catch clause.